firestore-bigquery-export/guides/GENERATE_SCHEMA_VIEWS.md

```diff
@@ -0,0 +1,374 @@
+    This view is analogous to the raw change-log, only it has typed columns
+    corresponding to fields of the schema.
+
++   `${param:PROJECT_ID}.${param:DATASET_ID}.${param:TABLE_ID}_schema_${SCHEMA_FILE_NAME}_latest`  
+  
+    This view stores typed rows for the documents currently in the collection.
+    This view is analogous to the "latest" view on the raw changelog, only it
+    includes the typed columns corresponding to fields in the corresponding
+    schema file.
+
+    Since `GEOGRAPHY` fields are not groupable entities in BigQuery (and the
+    query which builds the latest view of a collection of documents requires
+    grouping on the schema columns), the latest schema omits any `GEOGRAPHY`
+    columns and, instead, splits them out into two `NUMERIC` columns called
+    `${FIRESTORE_GEOPOINT}_latitude` and `${FIRESTORE_GEOPOINT}_longitude`.
+
+### Columns in a schema view
+
+Each schema view carries with it the following fields from the raw changelog:
+
++   `document_name STRING REQUIRED`
++   `timestamp TIMESTAMP REQUIRED`
++   `operation STRING REQUIRED`
+
+The remaining columns correspond to fields of the schema and are assigned types
+based on the corresponding Cloud Firestore types those fields have. With the
+exception of `map` and `array`, the type conversion scheme is as follows:
+
+<table>
+  <thead>
+    <tr>
+      <th><strong>Cloud Firestore type</strong></th>
+      <th><strong>BigQuery type</strong></th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>string</td>
+      <td>STRING</td>
+    </tr>
+    <tr>
+      <td>boolean</td>
+      <td>BOOLEAN</td>
+    </tr>
+    <tr>
+      <td>number</td>
+      <td>NUMERIC</td>
+    </tr>
+    <tr>
+      <td>timestamp</td>
+      <td>TIMESTAMP</td>
+    </tr>
+    <tr>
+      <td>geopoint</td>
+      <td>GEOGRAPHY</td>
+    </tr>
+    <tr>
+      <td>null</td>
+      <td>NULL</td>
+    </tr>
+  </tbody>
+</table>
+
+#### Cloud Firestore maps
+
+Cloud Firestore maps are interpreted recursively. If you include a map in your
+schema configuration, the resulting view will contain columns for whatever
+fields that map contains. If the map doesn't contain any fields, the map is
+ignored by the schema-views script.
+
+#### Cloud Firestore arrays
+
+Review [this example](https://github.com/firebase/extensions/blob/master/firestore-bigquery-export/guides/EXAMPLE_QUERIES.md) for querying an array.
+
+Cloud Firestore arrays are
+[unnested](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#unnest),
+so all array fields of the document are
+[cross joined](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#cross-join)
+in the output table. The view retains the member and offset columns, which are
+called `${FIRESTORE_ARRAY_NAME}_member` and `${FIRESTORE_ARRAY_NAME}_index`,
+respectively. To make querying easier, the view includes these two columns
+instead of the original `ARRAY` value field.
```

Comment:
It's probably worth pointing out that we don't support deeply nested arrays. Arrays are basically expanded to BigQuery `ARRAY<STRING>` where the contents the JSON-serialized contents of each member. We could expand this functionality in future work.

https://github.com/firebase/extensions/blob/next/firestore-bigquery-export/scripts/gen-schema-view/src/udf.ts#L69-L78

Reply:
To clarify, if I have an array that looks like this:

`[[1,2,3],[4,5,6]]`, it won't get double-unnested, it will just create two rows, one with `[1,2,3]` and one with `[4,5,6]`, correct?

Reply:
Per the [docs](https://firebase.google.com/docs/firestore/manage-data/data-types):

> An array cannot contain another array value as one of its elements.

Does that mean we don't need to worry about it?